spicy/src/compiler/codegen/parser-builder.cc

```diff
@@ -1045,6 +1045,17 @@ void ParserBuilder::newValueForField(const type::unit::item::Field& field, const
         nvalue = builder::member(state().self, field.id());
     }
 
+    if ( auto a = AttributeSet::find(field.attributes(), "&requires") ) {
+        // We evaluate "&requires" here so that the field's value has been
+        // set already, and is hence accessible to the condition through
+        // "self.<x>".
+        auto block = builder()->addBlock();
+        block->addLocal(ID("__dd"), field.parseType(), value);
+        auto cond = block->addTmp("requires", *a->valueAs<Expression>());
+        pushBuilder(block->addIf(builder::not_(cond)),
+                    [&]() { parseError("&required failed ($$ == %s)", {value}, a->value().location()); });
```

Comment:
We do not check any of the data dependencies of `cond` here and would trigger an `UnsetOptional` exception at runtime here if the condition accesses data which has not (yet) filled, e.g., in

```
type X = unit {
    a: uint8 &requires=(self.a > self.b); # self.b is always unset here.
    b: uint8;
};
```

While this makes sense, it can be nasty when e.g., reordering fields, when composing messages, or when making fields conditional with e.g., `switch`.

If we cannot validate that we only access `$$` and the final value of exactly this field we should at least add a test to ensure we trigger an exception.

Reply:
Maybe an overall better approach would be to evaluate these conditions only after all unit fields have been parsed. That would not only make the data dependency issue go away, but also allow users to check preconditions where they make sense (e.g., check that an optional field is only set if some other flag field has a certain value). A unit-level `&requires` could still make sense in that world to enforce even higher level constraints.

Reply:
The semantics here are similar to other field attributes (e.g., `&convert`), and I think it's also the right time to evaluate the constraint so that subsequent fields can rely on condition holding. 

I like the idea of a global `&requires`, I'll add that.